openedx_proversity_reports/middleware.py

```diff
@@ -0,0 +1,137 @@
+"""
+Middlewares for openedx-proversity-reports.
+"""
+import json
+from datetime import datetime, timedelta
+
+from django.conf import settings
+from openedx_proversity_reports.edxapp_wrapper.get_student_library import get_user_profile
+from openedx_proversity_reports.serializers import UserSessionSerializer
+
+
+class UserSessionMiddleware(object):
+    """
+    Middleware to store session data.
+    """
+    AVG_SESSION_LABEL = 'avg_session'
+    LAST_SESSION_LABEL = 'last_session'
+    LENGTH_CURRENT_SESSION_LABEL = 'length_current_session'
+    SESSION_NUMBER_LABEL = 'session_number'
+    TIME_BETWEEN_SESSIONS_LABEL = 'time_between_sessions'
+
+    def process_request(self, request):
+        """
+        This middleware stores the session date and calculates the
+        average session time, average time between sessions and current length
+        session for every authenticated user.
+        """
+        user = request.user
+
+        if user.is_anonymous():
+            return None
+
+        user_profile = get_user_profile().objects.get(user_id=user.id)
+
+        try:
+            meta = json.loads(user_profile.meta)
+            serializer = UserSessionSerializer(data=meta)
+            if serializer.is_valid():
+                session_data = serializer.validated_data
+            else:
+                session_data = {}
+
+        except ValueError:
+            session_data = {}
+
+        if not session_data:
+            user_profile.meta = self.get_meta_session_data(
+                avg_session=0,
+                length_current_session=0,
+                time_between_sessions=0,
+                session_number=1
+            )
+            user_profile.save()
+            return None
+
+        last_session = session_data.get(self.LAST_SESSION_LABEL)
+        avg_session = session_data.get(self.AVG_SESSION_LABEL, 0)
+        length_current_session = session_data.get(self.LENGTH_CURRENT_SESSION_LABEL, 0)
+        time_between_sessions = session_data.get(self.TIME_BETWEEN_SESSIONS_LABEL, 0)
+        session_number = session_data.get(self.SESSION_NUMBER_LABEL, 0)
+        delta_time = datetime.now(last_session.tzinfo) - last_session
```

Comment:
This will fail if last_session is None, please, add a default value for it. @andrey-canon

Reply:
If last_session is None, this line will never be executed